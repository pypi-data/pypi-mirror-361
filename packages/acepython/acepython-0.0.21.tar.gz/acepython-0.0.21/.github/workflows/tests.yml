name: Tests

on:
  pull_request:
    branches:
      - main


jobs:
  tests:
    name: ${{ matrix.session }} ${{ matrix.python }} / ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        include:
          - { python: "3.13", os: "ubuntu-latest", session: "tests" }
          - { python: "3.12", os: "ubuntu-latest", session: "tests" }
          - { python: "3.11", os: "ubuntu-latest", session: "tests" }
          - { python: "3.10", os: "ubuntu-latest", session: "tests" }

    env:
      NOXSESSION: ${{ matrix.session }}
      FORCE_COLOR: "1"
      PRE_COMMIT_COLOR: "always"

    steps:
      - name: Check out the repository
        uses: actions/checkout@v2.4.0

      - name: Set up Python ${{ matrix.python }}
        uses: actions/setup-python@v2.3.0
        with:
          python-version: ${{ matrix.python }}

      - name: Upgrade pip
        run: |
          pip install --upgrade pip virtualenv
          pip --version

      - name: Upgrade pip in virtual environments
        shell: python
        run: |
          import os
          import pip

          with open(os.environ["GITHUB_ENV"], mode="a") as io:
              print(f"VIRTUALENV_PIP={pip.__version__}", file=io)


      - name: Install Nox
        run: |
          pipx install nox
          nox --version


      - name: Run Nox
        run: |
          nox --force-color --python=${{ matrix.python }} --session=${{ matrix.session }}

