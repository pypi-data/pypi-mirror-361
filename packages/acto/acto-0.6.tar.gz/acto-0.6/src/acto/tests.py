from tclogger import shell_cmd, logger, get_now_str

from acto.periods import Perioder
from acto.retry import Retrier


def foo():
    cmd = 'date +"%T.%N"'
    shell_cmd(cmd, showcmd=False)


def foo_desc_func(x):
    func_strs = ['date +"%T.%N"']
    desc_str = f"foo at {x}"
    return func_strs, desc_str


def test_perioder():
    logger.note("> test_perioder")
    # patterns = "****-**-** **:**:**"
    patterns = {"second": "*[05]"}
    perioder = Perioder(patterns)
    perioder.bind(func=foo, desc_func=foo_desc_func)
    perioder.run()


foo_call_count = 0


def foo_to_retry(ok_until: int = 3):
    global foo_call_count
    foo_call_count += 1

    if foo_call_count < ok_until:
        raise RuntimeError(f"foo_to_retry failed at call: {foo_call_count}")
    else:
        cmd = 'date +"%T.%N"'
        shell_cmd(cmd, showcmd=False)


def test_retrier():
    logger.note("> test_retrier")
    with Retrier(max_retries=3, retry_interval=0.5) as retrier:
        retrier.run(foo_to_retry, ok_until=4)
    logger.mesg("* test_retrier completed")


if __name__ == "__main__":
    # test_perioder()
    test_retrier()

    # python -m acto.tests
