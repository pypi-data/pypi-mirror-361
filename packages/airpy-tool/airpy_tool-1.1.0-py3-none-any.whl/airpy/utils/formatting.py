"""
Utilities for formatting and cleaning dataframes.
"""
import pandas as pd
import numpy as np


def clean_dataframe(df: pd.DataFrame):
    """
    Clean a dataframe by removing units from column names.
    
    Args:
        df: Pandas DataFrame to clean
        
    Returns:
        Cleaned DataFrame
    """
    col_dict = {}
    for col in df.columns:
        if " (" in col:
            new_col = col.split(" (")[0]
            col_dict[col] = new_col
        else:
            col_dict[col] = col
    df.rename(columns=col_dict, inplace=True)
    return df


def read_df(df):
    """
    Clean and prepare a dataframe for processing.
    
    Args:
        df: Pandas DataFrame to process
        
    Returns:
        Processed DataFrame
    """
    # Check if 'Timestamp' or similar column exists and rename it to 'Timestamp'
    dd = ['Timestamp']
    for d in dd:
        if d in df.columns:
            df[d] = pd.to_datetime(df[d], errors='coerce', format='%Y-%m-%d %H:%M:%S')

    # Cleans all data for null entries and replaces with np.nan
    try:
        df['PM10'] = pd.to_numeric(df.PM10, errors='coerce')
    except:
        print("NO PM10 data")
        df['PM10'] = np.nan
    
    try:
        df['NO'] = pd.to_numeric(df.NO, errors='coerce')
    except:
        print("No NO data")
        df['NO'] = np.nan
    
    try:
        df['NO2'] = pd.to_numeric(df.NO2, errors='coerce')
    except:
        print("No NO2 data")
        df['NO2'] = np.nan
    
    try:
        df['NOx'] = pd.to_numeric(df.NOx, errors='coerce')
    except:
        print("NO NOx data")
        df['NOx'] = np.nan
    
    try:
        df['Ozone'] = pd.to_numeric(df.Ozone, errors='coerce')
    except:
        print("NO Ozone data")
        df['Ozone'] = np.nan
    
    try:
        df.rename(columns = {'PM2.5':'PM25'}, inplace = True)
        df['PM25'] = pd.to_numeric(df.PM25, errors='coerce')
    except:
        df['PM25'] = np.nan
        print("NO PM data")
    print("LOG: Standardized column names")
    return df


def get_formatted_df(path):
    """
    Function removes null entries and formats the date column.
    
    Args:
        path: Path to the data file
        
    Returns:
        Cleaned dataframe
    """
    if path.endswith('.csv'):
        df = pd.read_csv(path)
        df = clean_dataframe(df)
        df = read_df(df=df)
        df.drop(df.filter(regex="Unname"), axis=1, inplace=True)

    elif path.endswith('.xlsx'):
        df = pd.read_excel(path)
        df = df.iloc[16:].reset_index(drop=True)
        end_index = df[df.iloc[:, 0] == "Prescribed Standards"].index[0]
        df = df.iloc[:end_index-1].reset_index(drop=True)
        df.columns = ['From Date', 'To Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'SO2', 'CO',
                      'Ozone', 'Benzene', 'Toluene', 'Eth-Benzene', 'MP-Xylene', 'RH', 'WS', 'WD', 'SR',
                      'BP', 'Xylene', 'AT']
        df['Timestamp'] = df['From Date']
        print("Before cleaning: ", df.shape)
        df = clean_dataframe(df)
        df = read_df(df=df)
        df.drop(df.filter(regex="Unname"), axis=1, inplace=True)
    print("LOG: Dataframe cleaning completed")
    return df


def get_multiple_df_linerized(df1):
    """
    Function specific for CPCB outputs without any human intervention,
    removes null entries and formats the date column.
    
    Args:
        df1: Pandas DataFrame to process
        
    Returns:
        Tuple of (processed DataFrame, station_name, city, state)
    """
    from_year = df1['Unnamed: 1'][8][6:10]
    to_year = df1['Unnamed: 1'][9][6:10]
    station_name = df1['CENTRAL POLLUTION CONTROL BOARD'][11]
    lst = df1.index[df1['CENTRAL POLLUTION CONTROL BOARD'] == "From Date"].tolist()
    city = df1['Unnamed: 1'][4]
    state = df1['Unnamed: 1'][3]

    print("get_multiple_df_linerized")
    print(from_year, to_year, station_name, lst, city, state)

    count = 1
    for i in range(len(lst)):
        
        if (i+1 == len(lst)):
            df_temp = df1[lst[i]:].reset_index(drop=True)
        else:
            df_temp = df1[lst[i]:lst[i+1]-1].reset_index(drop=True)
        df_temp = df_temp.rename(columns=df_temp.iloc[0]).drop(df_temp.index[0])
        df_temp = df_temp.loc[:, df_temp.columns.notna()]
        
        if count != 1:
            del df_temp['From Date'], df_temp['To Date']
            df_concat = pd.concat([df_concat, df_temp], axis=1)
        else:
            df_concat = df_temp
        count = count + 1

    df_concat = df_concat.rename(columns = {'PM2.5':'PM25', 'From Date':'dates'})
    del df_concat['To Date']
    
    if(len(lst)>1):
        df_concat = df_concat[:len(df1[lst[0]:df1.index[df1['CENTRAL POLLUTION CONTROL BOARD'] == "Prescribed Standards"].tolist()[1]-1].reset_index(drop=True))-1]

    return df_concat, station_name, city, state 