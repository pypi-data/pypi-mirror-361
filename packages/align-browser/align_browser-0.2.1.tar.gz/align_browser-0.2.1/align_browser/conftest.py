#!/usr/bin/env python3
"""
Shared pytest fixtures for frontend testing.
"""

import json
import tempfile
import subprocess
import threading
import time
import yaml
import http.server
import socketserver
from pathlib import Path
from contextlib import contextmanager
import pytest
from playwright.sync_api import sync_playwright


class FrontendTestServer:
    """HTTP server for serving the built frontend during tests."""

    def __init__(self, dist_dir="dist", port=0):
        self.dist_dir = Path(dist_dir)
        self.port = port
        self.actual_port = None
        self.base_url = None
        self.server = None
        self.server_thread = None

    @contextmanager
    def run(self):
        """Context manager for running the test server."""

        class QuietHandler(http.server.SimpleHTTPRequestHandler):
            def log_message(self, format, *args):
                pass  # Suppress logging

        original_cwd = Path.cwd()

        try:
            # Change to dist directory
            if self.dist_dir.exists():
                import os

                os.chdir(self.dist_dir)

            # Start server in background thread
            class ReusableTCPServer(socketserver.TCPServer):
                allow_reuse_address = True

            with ReusableTCPServer(("", self.port), QuietHandler) as httpd:
                self.server = httpd
                self.actual_port = httpd.server_address[1]
                self.base_url = f"http://localhost:{self.actual_port}"

                self.server_thread = threading.Thread(
                    target=httpd.serve_forever, daemon=True
                )
                self.server_thread.start()

                # Wait for server to be ready
                time.sleep(0.5)

                yield self.base_url

        finally:
            # Restore original directory
            import os

            os.chdir(original_cwd)

            if self.server:
                self.server.shutdown()


class TestDataGenerator:
    """Generate minimal test data for frontend development."""

    @staticmethod
    def create_test_experiments():
        """Create test experiment data."""
        temp_dir = Path(tempfile.mkdtemp())
        experiments_root = temp_dir / "experiments"

        # Create realistic test experiments that match manifest structure
        test_configs = [
            # pipeline_baseline with Mistral (supports multiple KDMAs)
            {
                "adm_type": "pipeline_baseline",
                "llm": "mistralai/Mistral-7B-Instruct-v0.3",
                "kdmas": [
                    {"kdma": "affiliation", "value": 0.5},
                    {"kdma": "merit", "value": 0.7},
                ],
                "scenario": "test_scenario_1",
            },
            # Single KDMA experiments for test_scenario_1 (to support individual KDMA selection)
            {
                "adm_type": "pipeline_baseline",
                "llm": "mistralai/Mistral-7B-Instruct-v0.3",
                "kdmas": [{"kdma": "personal_safety", "value": 0.8}],
                "scenario": "test_scenario_1",
            },
            # Different scenario experiments
            {
                "adm_type": "pipeline_baseline",
                "llm": "mistralai/Mistral-7B-Instruct-v0.3",
                "kdmas": [{"kdma": "personal_safety", "value": 0.3}],
                "scenario": "test_scenario_2",
            },
            # Add pipeline_random experiments for scenario filtering tests
            {
                "adm_type": "pipeline_random",
                "llm": "no_llm",
                "kdmas": [],
                "scenario": "test_scenario_5",
            },
        ]

        for i, config in enumerate(test_configs):
            # Create pipeline directory structure
            pipeline_dir = experiments_root / config["adm_type"]
            pipeline_dir.mkdir(parents=True, exist_ok=True)

            # Create experiment directory with KDMA values in name
            kdma_parts = [f"{kdma['kdma']}-{kdma['value']}" for kdma in config["kdmas"]]
            exp_name = "_".join(kdma_parts) if kdma_parts else "no_kdma"
            experiment_dir = pipeline_dir / exp_name
            experiment_dir.mkdir(exist_ok=True)

            # Create .hydra directory and config.yaml
            hydra_dir = experiment_dir / ".hydra"
            hydra_dir.mkdir(exist_ok=True)

            # Create hydra config.yaml (this is what the parser expects)
            hydra_config = {
                "name": "test_experiment",
                "adm": {"name": config["adm_type"]},
                "alignment_target": {
                    "id": f"test-{i}",
                    "kdma_values": config["kdmas"],
                },
            }

            # Add LLM config if not no_llm
            if config["llm"] != "no_llm":
                hydra_config["adm"]["structured_inference_engine"] = {
                    "model_name": config["llm"]
                }

            with open(hydra_dir / "config.yaml", "w") as f:
                yaml.dump(hydra_config, f)

            # Create input/output data as single object (what Pydantic expects)
            input_output = {
                "input": {
                    "scenario_id": config["scenario"],
                    "state": f"Test scenario {config['scenario']} with medical triage situation",
                    "choices": [
                        {
                            "action_id": "action_a",
                            "kdma_association": {
                                kdma["kdma"]: 0.8 for kdma in config["kdmas"]
                            }
                            if config["kdmas"]
                            else {},
                            "unstructured": f"Take action A in {config['scenario']} - apply treatment",
                        },
                        {
                            "action_id": "action_b",
                            "kdma_association": {
                                kdma["kdma"]: 0.2 for kdma in config["kdmas"]
                            }
                            if config["kdmas"]
                            else {},
                            "unstructured": f"Take action B in {config['scenario']} - tag and evacuate",
                        },
                    ],
                },
                "output": {
                    "choice": "action_a",
                    "justification": f"Test justification for {config['scenario']}: This action aligns with the specified KDMA values.",
                },
            }

            with open(experiment_dir / "input_output.json", "w") as f:
                json.dump(input_output, f, indent=2)

            # Create scores file as single object (what Pydantic expects)
            scores = {
                "test_score": 0.85 + (i * 0.05),
                "scenario_id": config["scenario"],
            }

            with open(experiment_dir / "scores.json", "w") as f:
                json.dump(scores, f, indent=2)

            # Create timing file as single object (what Pydantic expects)
            timing = {
                "probe_time": 1234 + (i * 100),
                "scenario_id": config["scenario"],
            }

            with open(experiment_dir / "timing.json", "w") as f:
                json.dump(timing, f, indent=2)

        return experiments_root


@pytest.fixture(scope="session")
def built_frontend():
    """Use the existing built frontend for all tests."""
    # Use the existing dist directory that's already built
    project_root = Path(__file__).parent.parent
    dist_dir = project_root / "dist"

    # Ensure the dist directory exists and has the required files
    if not dist_dir.exists() or not (dist_dir / "manifest.json").exists():
        # Build the frontend if it doesn't exist
        cmd = ["uv", "run", "align-browser", "../experiments", "--dev", "--build-only"]
        result = subprocess.run(
            cmd, capture_output=True, text=True, cwd=str(project_root)
        )

        if result.returncode != 0:
            pytest.fail(f"Frontend build failed: {result.stderr}")

    yield dist_dir


@pytest.fixture(scope="session")
def test_server(built_frontend):
    """Provide a running test server."""
    server = FrontendTestServer(built_frontend, port=0)  # Use any available port
    with server.run() as base_url:
        yield base_url


@pytest.fixture(scope="session")
def browser_context():
    """Provide a browser context."""
    with sync_playwright() as p:
        browser = p.chromium.launch()
        context = browser.new_context()
        yield context
        context.close()
        browser.close()


@pytest.fixture
def page(browser_context):
    """Provide a browser page."""
    page = browser_context.new_page()
    yield page
    page.close()
