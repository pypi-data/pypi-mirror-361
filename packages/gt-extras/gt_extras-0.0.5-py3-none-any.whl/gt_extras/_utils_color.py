from __future__ import annotations

from great_tables._data_color.palettes import GradientPalette
from great_tables._data_color.constants import DEFAULT_PALETTE, ALL_PALETTES
from great_tables._data_color.base import (
    _html_color,
    _rescale_factor,
    _get_domain_factor,
)

from great_tables._tbl_data import TblData

__all__ = ["_get_discrete_colors_from_palette"]


def _get_discrete_colors_from_palette(
    palette: list[str] | str | None, data: list, data_table: TblData
) -> list[str]:
    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    domain_factor = _get_domain_factor(df=data_table, vals=data)
    scaled_vals = _rescale_factor(
        df=data_table, vals=data, domain=domain_factor, palette=palette
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_vals)

    return color_vals
