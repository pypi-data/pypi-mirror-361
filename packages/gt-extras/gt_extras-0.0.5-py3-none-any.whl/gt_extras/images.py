from __future__ import annotations

from great_tables import html
from great_tables._text import Html

__all__ = ["img_header"]


def img_header(
    label: str,
    img_url: str,
    height: float = 60,
    font_size: int = 12,
    border_color: str = "black",
    text_color: str = "black",
) -> Html:
    """
    Create an HTML header with an image and a label, apt for a column label.

    Parameters
    ----------
    label
        The text label to display below the image.

    img_url
        The URL of the image to display. This can be a filepath or an image on the web.

    height
        The height of the image in pixels.

    font_size
        The font size of the label text.

    border_color
        The color of the border below the image.

    text_color
        The color of the label text.

    Returns
    -------
    html
        A Great Tables `html` element for the header.

    Examples
    -------
    ```{python}
    import pandas as pd
    from great_tables import GT, md
    import gt_extras as gte

    df = pd.DataFrame(
        {
            "Category": ["Points", "Rebounds", "Assists", "Blocks", "Steals"],
            "Hart": [1051, 737, 453, 27, 119],
            "Brunson": [1690, 187, 475, 8, 60],
            "Bridges": [1444, 259, 306, 43, 75],
        }
    )

    hart_header = gte.img_header(
        label="Josh Hart",
        img_url="https://a.espncdn.com/combiner/i?img=/i/headshots/nba/players/full/3062679.png",
    )

    brunson_header = gte.img_header(
        label="Jalen Brunson",
        img_url="https://a.espncdn.com/combiner/i?img=/i/headshots/nba/players/full/3934672.png",
    )

    bridges_header = gte.img_header(
        label="Mikal Bridges",
        img_url="https://a.espncdn.com/combiner/i?img=/i/headshots/nba/players/full/3147657.png",
    )

    (
        GT(df, rowname_col="Category")
        .tab_source_note(md("Images and data courtesy of [ESPN](https://www.espn.com)"))
        .cols_label(
            {
                "Hart": hart_header,
                "Brunson": brunson_header,
                "Bridges": bridges_header,
            }
        )
    )
    ```
    """

    img_html = f"""
    <img src="{img_url}" style="
        height:{height}px;
        border-bottom:2px solid {border_color};"
    />
    """.strip()

    label_html = f"""
    <div style="
        font-size:{font_size}px;
        color:{text_color};
        text-align:center;
        width:100%;
    ">
        {label}
    </div>
    """.strip()

    full_element = f"""
    <div style="text-align:center;">
        {img_html}
        {label_html}
    </div>
    """.strip()

    return html(full_element)
