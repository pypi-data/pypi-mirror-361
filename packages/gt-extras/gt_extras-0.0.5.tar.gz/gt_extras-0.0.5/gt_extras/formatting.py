from __future__ import annotations

import pandas as pd
import polars as pl
from great_tables import GT
from great_tables._gt_data import Boxhead, ColInfo
from great_tables._tbl_data import SelectExpr, copy_frame, is_na

from gt_extras._utils_column import _validate_and_get_single_column

__all__ = ["fmt_pct_extra", "gt_duplicate_column"]


def fmt_pct_extra(
    gt: GT,
    columns: SelectExpr,
    scale: float = 100,
    threshold: float = 1.0,
    color: str = "grey",
    decimals: int = 1,
) -> GT:
    """
    Convert to percent and show less than `1%` as `<1%` in grey.

    The `fmt_pct_extra()` function takes an existing `GT` object and formats a column of numeric
    values as percentages. Values below the specified threshold are displayed as `"<threshold%"`
    instead of their actual percentage value, and in a unique color.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    columns
        The columns containing numeric values to format as percentages.

    scale
        Multiplication factor to convert values to percentages.
        Use `100` if values are decimals `(0.05 -> 5%)` (default),
        use `1` if values are already percentages `(5 -> 5%)`.

    threshold
        The percentage threshold below which values are displayed as `"<threshold%"` instead of
        their actual value. Note this refers to the scaled value, not the original.

    color
        The color to use for values below the threshold.

    decimals
        Number of decimal places to display for percentages.

    Returns
    -------
    GT
        A `GT` object with formatted percentage column.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import towny
    import gt_extras as gte

    towny_mini = towny[
        [
            "name",
            "pop_change_1996_2001_pct",
            "pop_change_2001_2006_pct",
            "pop_change_2006_2011_pct",
        ]
    ].tail(10)

    gt = (
        GT(towny_mini)
        .tab_spanner(label="Population Change", columns=[1, 2, 3])
        .cols_label(
            pop_change_1996_2001_pct="'96-'01",
            pop_change_2001_2006_pct="'01-'06",
            pop_change_2006_2011_pct="'06-'11",
        )
    )

    gt.pipe(
        gte.fmt_pct_extra,
        columns=[1, 2, 3],
        threshold=5,
        color="green",
    )
    ```
    """
    # TODO: consider how to handle negative values

    def _fmt_pct_single_val(value: float):
        if is_na(gt._tbl_data, value):
            return ""

        # Convert to percentage
        pct_value = value * scale

        if abs(pct_value) < threshold:
            return f"<span style='color:{color};'><{threshold:g}%</span>"
        else:
            return f"{pct_value:.{decimals}f}%"

    res = gt
    res = res.fmt(_fmt_pct_single_val, columns=columns)

    return res


def gt_duplicate_column(
    gt: GT,
    column: SelectExpr,
    after: str | None = None,
    append_text: str | None = "_dupe",
    dupe_name: str | None = None,
) -> GT:
    """
    Duplicate a column in a `GT` object.

    The `gt_duplicate_column()` function takes an existing `GT` object and creates a duplicate
    (without styling) of the specified column. The duplicated column can be renamed using either
    `dupe_name` or by appending text to the original column name, and positioned at a
    specific location in the table.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    column
        The column to duplicate. Can be a column name or index.

    after
        The column after which to place the duplicated column. If `None`, the duplicated
        column will be moved to the end of the table.

    append_text
        Text to append to the original column name for the duplicate. Only used if
        `dupe_name` is not provided. Defaults to `"_dupe"`.

    dupe_name
        The name for the duplicated column. If provided, this overrides `append_text`.

    Returns
    -------
    GT
        A `GT` object with the duplicated column added.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["mfr", "model", "year", "hp"]].head(5)
    gt = GT(gtcars_mini)

    # Duplicate with custom name and position
    gt.pipe(
        gte.gt_duplicate_column,
        column="hp",
        after="year",
    )
    ```
    """
    original_name, _ = _validate_and_get_single_column(gt, column)

    # If dupe_name is given, it overrides append_text
    append_text = append_text or "_dupe"
    if dupe_name is not None:
        new_col_name = dupe_name
    else:
        new_col_name = original_name + append_text

    if new_col_name == original_name:
        raise ValueError(
            f"The new column name '{new_col_name}' cannot be the same as the original column name '{original_name}'."
        )

    res = gt
    new_data_table = copy_frame(res._tbl_data)
    new_body = res._body.copy()

    # get the boxhead info
    original_col_info = None
    for col_info in res._boxhead:
        if col_info.var == original_name:
            original_col_info = col_info
            break

    # make the new boxhead entry
    new_col_info = ColInfo(
        var=new_col_name,
        type=original_col_info.type,
        column_label=new_col_name,
        column_align=original_col_info.column_align,
        column_width=original_col_info.column_width,
    )

    # A little clunky, but I dont have any other solutions
    if isinstance(new_data_table, pd.DataFrame):
        new_data_table[new_col_name] = new_data_table[original_name]
        new_body.body[new_col_name] = new_body.body[original_name]

    elif isinstance(new_data_table, pl.DataFrame):
        new_data_table = new_data_table.with_columns(
            new_data_table[original_name].alias(new_col_name)
        )
        new_body.body = new_body.body.with_columns(
            new_body.body[original_name].alias(new_col_name)
        )

    else:
        raise TypeError(
            """Unsupported type.
            This function will only work if the underlying data is a Polars or Pandas dataframe."""
        )

    new_boxhead_list = list(res._boxhead._d) + [new_col_info]
    new_boxhead = Boxhead(new_boxhead_list)

    res = res._replace(_tbl_data=new_data_table, _boxhead=new_boxhead, _body=new_body)

    if after is None:
        res = res.cols_move_to_end(new_col_name)
    else:
        res = res.cols_move(new_col_name, after=after)

    return res
