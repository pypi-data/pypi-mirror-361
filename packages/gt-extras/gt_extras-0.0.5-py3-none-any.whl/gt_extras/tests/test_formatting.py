import numpy as np
import pandas as pd
import polars as pl
import pyarrow as pa
import pytest
from great_tables import GT

from gt_extras.formatting import fmt_pct_extra, gt_duplicate_column
from gt_extras.tests.conftest import assert_rendered_body


def test_fmt_pct_extra_snap(snapshot, mini_gt):
    res = fmt_pct_extra(mini_gt, columns="num", scale=1, decimals=0)
    assert_rendered_body(snapshot, gt=res)


def test_fmt_pct_extra_basic(mini_gt):
    html = fmt_pct_extra(mini_gt, columns="num", scale=1).as_raw_html()

    assert "<span style='color:grey;'><1%</span>" in html
    assert "2.2%" in html
    assert "33.3%" in html


def test_fmt_pct_extra_threshold_low(mini_gt):
    html = fmt_pct_extra(mini_gt, columns="num", scale=100, threshold=10).as_raw_html()

    assert "11.1%" in html
    assert "222.2%" in html
    assert "3333.0%" in html


def test_fmt_pct_extra_threshold_high(mini_gt):
    html = fmt_pct_extra(
        mini_gt, columns="num", scale=100, threshold=4000
    ).as_raw_html()

    assert html.count("<span style='color:grey;'><4000%</span>") == 3


def test_fmt_pct_extra_custom_color(mini_gt):
    html = fmt_pct_extra(
        mini_gt, columns="num", color="red", threshold=50
    ).as_raw_html()

    assert html.count("<span style='color:red;'><50%</span>") == 1


def test_fmt_pct_extra_decimals(mini_gt):
    html_0 = fmt_pct_extra(mini_gt, columns="num", decimals=0).as_raw_html()
    assert "11%" in html_0  # 0.1111 * 100 = 11.11% rounded
    assert "222%" in html_0


def test_fmt_pct_extra_negative_values():
    df = pd.DataFrame({"num": [-0.005, -0.25, 0.15]})
    gt_test = GT(df)

    html = fmt_pct_extra(gt=gt_test, columns="num", threshold=1.0).as_raw_html()

    assert "<span style='color:grey;'><1%</span>" in html
    assert "-25.0%" in html
    assert "15.0%" in html


def test_fmt_pct_extra_zero_values():
    df = pd.DataFrame({"num": [0.0, 0.005, 0.02]})
    gt_test = GT(df)
    html = fmt_pct_extra(gt=gt_test, columns="num").as_raw_html()

    assert html.count("<span style='color:grey;'><1%</span>") == 2
    assert "2.0%" in html


def test_fmt_pct_extra_edge_case_threshold():
    df = pd.DataFrame({"num": [0.01, 0.0099, 0.0101]})
    gt_test = GT(df)

    html = fmt_pct_extra(
        gt=gt_test, columns="num", scale=100, threshold=1.0, decimals=2
    ).as_raw_html()

    assert "1.00%" in html
    assert "<span style='color:grey;'><1%</span>" in html
    assert "1.01%" in html


def test_fmt_pct_extra_with_none_values():
    df = pd.DataFrame({"num": [0.005, None, 0.25, np.nan]})
    gt_test = GT(df)

    result = fmt_pct_extra(gt=gt_test, columns="num")
    html = result.as_raw_html()

    assert isinstance(result, GT)
    assert "25%" in html


def test_gt_duplicate_column_snap(snapshot, mini_gt):
    res = gt_duplicate_column(mini_gt, column="num")
    assert_rendered_body(snapshot, gt=res)


def test_gt_duplicate_column_basic(mini_gt):
    res = gt_duplicate_column(mini_gt, column="num", append_text="_copy")
    html = res.as_raw_html()

    assert "num_copy" in res._tbl_data.columns
    assert "num_copy" in html
    assert all(res._tbl_data["num"] == res._tbl_data["num_copy"])


def test_gt_duplicate_column_custom_name(mini_gt):
    res = gt_duplicate_column(mini_gt, column="num", dupe_name="duplicated_num")
    html = res.as_raw_html()

    assert "duplicated_num" in res._tbl_data.columns
    assert "duplicated_num" in html
    assert all(res._tbl_data["num"] == res._tbl_data["duplicated_num"])


def test_gt_duplicate_column_position(mini_gt):
    res = gt_duplicate_column(mini_gt, column="num", after="char")
    html = res.as_raw_html()

    assert "num_dupe" in res._tbl_data.columns
    assert "num_dupe" in html

    columns = list(res._boxhead)
    assert columns[2].column_label == "num_dupe"


def test_gt_duplicate_column_polars():
    df = pl.DataFrame({"num": [1, 2, 3], "char": ["a", "b", "c"]})
    gt_test = GT(df)

    res = gt_duplicate_column(gt_test, column="num", append_text="_copy")
    html = res.as_raw_html()

    assert "num_copy" in res._tbl_data.columns
    assert "num_copy" in html

    original_values = res._tbl_data.get_column("num").to_list()
    duplicated_values = res._tbl_data.get_column("num_copy").to_list()
    assert original_values == duplicated_values


def test_gt_duplicate_column_invalid_type():
    data = pa.table({"num": [1, 2, 3], "char": ["a", "b", "c"]})
    gt_test = GT(data)

    with pytest.raises(TypeError, match="Unsupported type"):
        gt_duplicate_column(gt_test, column="num")


def test_gt_duplicate_column_invalid_name(mini_gt):
    with pytest.raises(
        ValueError, match="cannot be the same as the original column name"
    ):
        gt_duplicate_column(mini_gt, column="num", dupe_name="num")
