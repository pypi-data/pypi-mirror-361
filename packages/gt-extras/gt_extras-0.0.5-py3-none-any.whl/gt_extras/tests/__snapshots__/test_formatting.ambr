# serializer version: 1
# name: test_fmt_pct_extra_snap
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right"><span style='color:grey;'><1%</span></td>
      <td class="gt_row gt_left">apricot</td>
      <td class="gt_row gt_left">one</td>
      <td class="gt_row gt_right">2015-01-15</td>
      <td class="gt_row gt_right">13:35</td>
      <td class="gt_row gt_right">2018-01-01 02:22</td>
      <td class="gt_row gt_right">49.95</td>
      <td class="gt_row gt_left">row_1</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2%</td>
      <td class="gt_row gt_left">banana</td>
      <td class="gt_row gt_left">two</td>
      <td class="gt_row gt_right">2015-02-15</td>
      <td class="gt_row gt_right">14:40</td>
      <td class="gt_row gt_right">2018-02-02 14:33</td>
      <td class="gt_row gt_right">17.95</td>
      <td class="gt_row gt_left">row_2</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33%</td>
      <td class="gt_row gt_left">coconut</td>
      <td class="gt_row gt_left">three</td>
      <td class="gt_row gt_right">2015-03-15</td>
      <td class="gt_row gt_right">15:45</td>
      <td class="gt_row gt_right">2018-03-03 03:44</td>
      <td class="gt_row gt_right">1.39</td>
      <td class="gt_row gt_left">row_3</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_gt_duplicate_column_snap
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
      <td class="gt_row gt_left">one</td>
      <td class="gt_row gt_right">2015-01-15</td>
      <td class="gt_row gt_right">13:35</td>
      <td class="gt_row gt_right">2018-01-01 02:22</td>
      <td class="gt_row gt_right">49.95</td>
      <td class="gt_row gt_left">row_1</td>
      <td class="gt_row gt_left">grp_a</td>
      <td class="gt_row gt_right">0.1111</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
      <td class="gt_row gt_left">two</td>
      <td class="gt_row gt_right">2015-02-15</td>
      <td class="gt_row gt_right">14:40</td>
      <td class="gt_row gt_right">2018-02-02 14:33</td>
      <td class="gt_row gt_right">17.95</td>
      <td class="gt_row gt_left">row_2</td>
      <td class="gt_row gt_left">grp_a</td>
      <td class="gt_row gt_right">2.222</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
      <td class="gt_row gt_left">three</td>
      <td class="gt_row gt_right">2015-03-15</td>
      <td class="gt_row gt_right">15:45</td>
      <td class="gt_row gt_right">2018-03-03 03:44</td>
      <td class="gt_row gt_right">1.39</td>
      <td class="gt_row gt_left">row_3</td>
      <td class="gt_row gt_left">grp_a</td>
      <td class="gt_row gt_right">33.33</td>
    </tr>
  </tbody>
  '''
# ---
